### PR TITLE
Fix send/sendto

### PR DESCRIPTION
**Pull request by kwagyeman**
Both methods can move megabytes of data now without issues.

Also, fixed an uninitialized value returning from recv_from when you poll
for data and there's nothing to get.

### CODE REVIEW

File: src/winc1500/src/winc.c [hunk 1]
```diff
@@ -820,23 +822,33 @@ int winc_socket_connect(int fd, sockaddr *addr, uint32_t timeout)
...
             // Do async request
-            ret = winc_async_request(SOCKET_MSG_SEND, &n, timeout);
+            int async_ret;
+            ret = winc_async_request(SOCKET_MSG_SEND, &async_ret, timeout);
 
             // Check sent bytes returned from async request.
-            if (ret != SOCK_ERR_NO_ERROR || n <= 0) {
-                return (n <= 0)? n : ret;
-            }
+            if (ret != SOCK_ERR_NO_ERROR) return ret;
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:840]: This is basically the same as the old code. There's no reason to change it (except for n<0), and please reuse n.

File: src/winc1500/src/winc.c [hunk 2]
```diff
@@ -820,23 +822,33 @@ int winc_socket_connect(int fd, sockaddr *addr, uint32_t timeout)
...
+            if (async_ret < 0) return async_ret;
+
+            bytes += async_ret;
+        } else if (ret != SOCK_ERR_BUFFER_FULL) {
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:844]: Just a small style change:
```
if (ret == SOCK_ERR_NO_ERROR) {
   /// do the async call
} else if (ret == SOCK_ERR_BUFFER_FULL) {
 // wait until timeout
} else { // another error
  return ret;
}
```

File: src/winc1500/src/winc.c [hunk 3]
```diff
@@ -856,9 +868,8 @@ int winc_socket_recv(int fd, uint8_t *buf, uint32_t len, winc_socket_buf_t *sock
             ret = winc_async_request(SOCKET_MSG_RECV, &recv_bytes, timeout);
         }
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:870]: This is exactly the same as the removed code, so why is this change needed ? Unless you want to check if winc_recv failed first (it's not likely to happen), but you could just check ret first:
```
if (ret != SOCK_ERR_NO_ERROR || recv_bytes <= 0) {
     return (ret != SOCK_ERR_NO_ERROR)? ret: recv_bytes;
}
```

File: src/winc1500/src/winc.c [hunk 4]
```diff
@@ -872,27 +883,51 @@ int winc_socket_recv(int fd, uint8_t *buf, uint32_t len, winc_socket_buf_t *sock
 
 int winc_socket_sendto(int fd, const uint8_t *buf, uint32_t len, sockaddr *addr, uint32_t timeout)
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:884]: Same comments as send.

File: src/winc1500/src/winc.c [hunk 5]
```diff
@@ -872,27 +883,51 @@ int winc_socket_recv(int fd, uint8_t *buf, uint32_t len, winc_socket_buf_t *sock
...
 int winc_socket_recvfrom(int fd, uint8_t *buf, uint32_t len, sockaddr *addr, uint32_t timeout)
 {
+    memset(addr, 0, sizeof(sockaddr));
+
     recv_from_t rfrom;
     int ret = WINC1500_EXPORT(recvfrom)(fd, buf, len, timeout);
     if (ret == SOCK_ERR_NO_ERROR) {
         // Do async request
         ret = winc_async_request(SOCKET_MSG_RECVFROM, &rfrom, timeout);
     }
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:928]: Please keep the old code and switch ret and rfrom.size.

File: src/winc1500/src/winc.c (hunk 1 above)
Comment (kwagyeman) [reply src/winc1500/src/winc.c:840]: I can't do this for all cases because sendto returns 0 for n always. For consistency I don't want to do this.

File: src/winc1500/src/winc.c [hunk 6]
```diff
@@ -820,23 +822,36 @@ int winc_socket_connect(int fd, sockaddr *addr, uint32_t timeout)
...
             // Check sent bytes returned from async request.
-            if (ret != SOCK_ERR_NO_ERROR || n <= 0) {
-                return (n <= 0)? n : ret;
+            if (ret != SOCK_ERR_NO_ERROR || async_ret <= 0) {
+                return (ret != SOCK_ERR_NO_ERROR) ? ret : async_ret;
             }
+
+            bytes += async_ret;
+        } else if (ret == SOCK_ERR_BUFFER_FULL) {
+            // wait until timeout
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:846]: The timeout should be here.

File: src/winc1500/src/winc.c [hunk 7]
```diff
@@ -820,23 +822,36 @@ int winc_socket_connect(int fd, sockaddr *addr, uint32_t timeout)
...
+        } else { // another error
+            return ret;
+        }
+
+        // timeout == 0 in blocking mode.
+        if (timeout && ((HAL_GetTick() - tick_start) >= timeout)) {
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:852]: This runs in every iteration of the loop.

File: src/winc1500/src/winc.c [hunk 8]
```diff
@@ -872,26 +888,55 @@ int winc_socket_recv(int fd, uint8_t *buf, uint32_t len, winc_socket_buf_t *sock
...
+            // Do async request
+            int async_ret; // sendto always returns 0.
+            ret = winc_async_request(SOCKET_MSG_SENDTO, &async_ret, timeout);
+
+            // Check sent bytes returned from async request.
+            if (ret != SOCK_ERR_NO_ERROR || async_ret < 0) {
+                return (ret != SOCK_ERR_NO_ERROR) ? ret : async_ret;
+            }
+
+            bytes += n;
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:910]: bytes += async_ret.

File: src/winc1500/src/winc.c [hunk 9]
```diff
@@ -872,26 +888,55 @@ int winc_socket_recv(int fd, uint8_t *buf, uint32_t len, winc_socket_buf_t *sock
...
+        } else if (ret == SOCK_ERR_BUFFER_FULL) {
+            // wait until timeout
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:912]: timeout should be here.

File: src/winc1500/src/winc.c [hunk 10]
```diff
@@ -872,26 +888,55 @@ int winc_socket_recv(int fd, uint8_t *buf, uint32_t len, winc_socket_buf_t *sock
...
+        } else { // another error
+            return ret;
+        }
+
+        // timeout == 0 in blocking mode.
+        if (timeout && ((HAL_GetTick() - tick_start) >= timeout)) {
```
Comment (iabdalkader) [inline src/winc1500/src/winc.c:918]: This runs in every iteration of the loop.
